src/storage/src/hummock/sstable_store.rs

```diff
@@ -960,6 +1042,154 @@ impl BlockStream {
+        Self {
+            byte_stream,
+            block_idx: 0,
+            block_metas,
+            buf: Bytes::default(),
+            buff_offset: 0,
+            object_id,
+            cache,
+            blocks: VecDeque::default(),
+            start_block_index,
+        }
+    }
+
+    /// Reads the next block from the stream and returns it. Returns `None` if there are no blocks
+    /// left to read.
+    pub async fn next_block(&mut self) -> HummockResult<Option<BlockHolder>> {
+        if self.block_idx >= self.block_metas.len() {
+            return Ok(None);
+        }
+
+        let block_meta = &self.block_metas[self.block_idx];
+        fail_point!("stream_batch_read_err", |_| Err(
+            HummockError::object_io_error(ObjectError::internal("stream read error"))
+        ));
+        if let Some(block) = self.blocks.pop_front() {
+            self.block_idx += 1;
+            return Ok(Some(block));
+        }
+        let mut read_buf = BytesMut::with_capacity(block_meta.len as usize);
+        let start_pos = if self.buff_offset < self.buf.len() {
+            read_buf.extend_from_slice(&self.buf[self.buff_offset..]);
+            self.buf.len() - self.buff_offset
+        } else {
+            0
+        };
+        let mut rest = block_meta.len as usize - start_pos;
+        self.buff_offset = 0;
+        while rest > 0 {
+            let next_packet = match self.byte_stream.read_bytes().await {
+                Some(Ok(data)) => data,
+                Some(Err(e)) => return Err(e.into()),
+                None => {
+                    panic!("read eof: {}, {}", start_pos, rest);
+                }
+            };
+
+            let read_len = std::cmp::min(next_packet.len(), rest);
+            read_buf.extend_from_slice(&next_packet[..read_len]);
+            rest -= read_len;
+            if rest == 0 {
+                self.buff_offset = read_len;
+                self.buf = next_packet;
+                break;
+            }
+        }
+        let block = Block::decode(
+            read_buf.freeze(),
+            self.block_metas[self.block_idx].uncompressed_size as usize,
+        )?;
+        let holder = self.cache.insert(
+            self.object_id,
+            (self.start_block_index + self.block_idx) as u64,
+            Box::new(block),
+            CachePriority::Low,
+        );
+        let mut block_idx = self.block_idx + 1;
+        let mut buff_offset = self.buff_offset;
+        while block_idx < self.block_metas.len() {
+            let end = buff_offset + self.block_metas[block_idx].len as usize;
+            if end > self.buf.len() {
+                break;
+            }
+            // copy again to avoid hold a large bytes reference in block-cache.
+            let block = Block::decode(
+                Bytes::copy_from_slice(&self.buf[buff_offset..end]),
+                self.block_metas[block_idx].uncompressed_size as usize,
+            )?;
```

Comment:
While here access `self.buf` by buffer_offset. Why? It seems doesn't make sense.

Reply:
Emmm ? I do not understand what do you mean

Reply:
Never mind. I misunderstood previously.